semantic/constraints.go

```diff
@@ -306,14 +306,31 @@ func (v ConstraintGenerator) typeof(n Node) (PolyType, error) {
 		if !ok {
 			return nil, errors.New("member object must be a type variable")
 		}
-		v.cs.AddKindConst(tv, KRecord{
+		v.cs.AddKindConst(tv, ObjectKind{
 			properties: map[string]PolyType{n.Property: ptv},
 			lower:      LabelSet{n.Property},
 			upper:      AllLabels(),
 		})
 		return ptv, nil
+	case *IndexExpression:
+		ptv := v.cs.f.Fresh()
+		t, err := v.lookup(n.Array)
+		if err != nil {
+			return nil, err
+		}
+		tv, ok := t.(Tvar)
+		if !ok {
+			return nil, errors.New("array must be a type variable")
+		}
+		idx, err := v.lookup(n.Index)
+		if err != nil {
+			return nil, err
+		}
+		v.cs.AddKindConst(tv, ArrayKind{ptv})
+		v.cs.AddTypeConst(idx, Int, n.Index.Location())
+		return ptv, nil
 	case *ArrayExpression:
-		at := array{typ: NewObjectPolyType(nil, nil, AllLabels())}
+		at := array{typ: v.cs.f.Fresh()}
```

Comment:
Using a type variable makes a lot more sense :), With this change the loop below does not need to be inside a len() > 0 check and does not need to treat the first element as special. Instead the loop can now just unify against `at.typ` directly. Can you make that simplification?